// typed-mite-api/dist/lib/mite-responses.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.MiteServerErrorResponse = exports.MiteLockedResponse = exports.MiteUnprocessableEntityResponse = exports.MiteNotAcceptableResponse = exports.MiteNotFoundResponse = exports.MiteForbiddenResponse = exports.MiteUnauthorizedResponse = exports.MiteBadRequestResponse = exports.MiteCreatedResponse = exports.MiteSuccessResponse = exports.MiteResponse = void 0;
var MiteResponse = /** @class */ (function (_super) {
    __extends(MiteResponse, _super);
    function MiteResponse(name, message) {
        var _this = _super.call(this, message) || this;
        _this.name = name;
        _this.message = message;
        return _this;
    }
    return MiteResponse;
}(Error));
exports.MiteResponse = MiteResponse;
// Response code 200
var MiteSuccessResponse = /** @class */ (function (_super) {
    __extends(MiteSuccessResponse, _super);
    function MiteSuccessResponse(message) {
        return _super.call(this, "MiteSuccessResponse", message) || this;
    }
    return MiteSuccessResponse;
}(MiteResponse));
exports.MiteSuccessResponse = MiteSuccessResponse;
// Response code 201
var MiteCreatedResponse = /** @class */ (function (_super) {
    __extends(MiteCreatedResponse, _super);
    function MiteCreatedResponse(message) {
        return _super.call(this, "MiteCreatedResponse", message) || this;
    }
    return MiteCreatedResponse;
}(MiteResponse));
exports.MiteCreatedResponse = MiteCreatedResponse;
// Response code 400
var MiteBadRequestResponse = /** @class */ (function (_super) {
    __extends(MiteBadRequestResponse, _super);
    function MiteBadRequestResponse(message) {
        return _super.call(this, "MiteBadRequestResponse", message) || this;
    }
    return MiteBadRequestResponse;
}(MiteResponse));
exports.MiteBadRequestResponse = MiteBadRequestResponse;
// Response code 401
var MiteUnauthorizedResponse = /** @class */ (function (_super) {
    __extends(MiteUnauthorizedResponse, _super);
    function MiteUnauthorizedResponse(message) {
        return _super.call(this, "MiteUnauthorizedResponse", message) || this;
    }
    return MiteUnauthorizedResponse;
}(MiteResponse));
exports.MiteUnauthorizedResponse = MiteUnauthorizedResponse;
// Response code 403
var MiteForbiddenResponse = /** @class */ (function (_super) {
    __extends(MiteForbiddenResponse, _super);
    function MiteForbiddenResponse(message) {
        return _super.call(this, "MiteForbiddenResponse", message) || this;
    }
    return MiteForbiddenResponse;
}(MiteResponse));
exports.MiteForbiddenResponse = MiteForbiddenResponse;
// Response code 404
var MiteNotFoundResponse = /** @class */ (function (_super) {
    __extends(MiteNotFoundResponse, _super);
    function MiteNotFoundResponse(message) {
        return _super.call(this, "MiteNotFoundResponse", message) || this;
    }
    return MiteNotFoundResponse;
}(MiteResponse));
exports.MiteNotFoundResponse = MiteNotFoundResponse;
// Response code 406
var MiteNotAcceptableResponse = /** @class */ (function (_super) {
    __extends(MiteNotAcceptableResponse, _super);
    function MiteNotAcceptableResponse(message) {
        return _super.call(this, "MiteNotAcceptableResponse", message) || this;
    }
    return MiteNotAcceptableResponse;
}(MiteResponse));
exports.MiteNotAcceptableResponse = MiteNotAcceptableResponse;
// Response code 422
var MiteUnprocessableEntityResponse = /** @class */ (function (_super) {
    __extends(MiteUnprocessableEntityResponse, _super);
    function MiteUnprocessableEntityResponse(message) {
        return _super.call(this, "MiteUnprocessableEntityResponse", message) || this;
    }
    return MiteUnprocessableEntityResponse;
}(MiteResponse));
exports.MiteUnprocessableEntityResponse = MiteUnprocessableEntityResponse;
// Response code 423
var MiteLockedResponse = /** @class */ (function (_super) {
    __extends(MiteLockedResponse, _super);
    function MiteLockedResponse(message) {
        return _super.call(this, "MiteLockedResponse", message) || this;
    }
    return MiteLockedResponse;
}(MiteResponse));
exports.MiteLockedResponse = MiteLockedResponse;
// Response codes 500, 502, 503 Server Error
var MiteServerErrorResponse = /** @class */ (function (_super) {
    __extends(MiteServerErrorResponse, _super);
    function MiteServerErrorResponse(message) {
        return _super.call(this, "MiteServerErrorResponse", message) || this;
    }
    return MiteServerErrorResponse;
}(MiteResponse));
exports.MiteServerErrorResponse = MiteServerErrorResponse;
